### PR TITLE
wg_engine: Introduce ClipPath composition support

### PR DESCRIPTION
**Pull request by SergeyLebedkin**
[issues 1479: ClipPath](#1479)

Introduce ClipPath composition support.
Clip path composition is an only composition type who doesn't ignore blend method. 
Clip path is a combination of composition approach and blend approach using compute shader.

image

### CODE REVIEW

**Comment by mgrudzinska** [comment]
@SergeyLebedkin some issues:

1. clipping a shape that is later used as a clip 

```
        auto shape = tvg::Shape::gen();
        shape->appendCircle(125, 100, 150, 150);
        shape->fill(255, 255, 255, 200);

        auto clip1 = tvg::Shape::gen();
        clip1->appendCircle(125, 100, 50, 50);

        auto clip2 = tvg::Shape::gen();
        clip2->appendCircle(175, 100, 50, 15);
        clip1->composite(std::move(clip2), tvg::CompositeMethod::ClipPath); //ignored in wg?
        shape->composite(std::move(clip1), tvg::CompositeMethod::ClipPath);
        canvas->push(std::move(shape));
```
Zrzut ekranu 2024-07-14 o 01 02 44

2. please double check SceneClipper.cpp example (probably another pr?):
`Assertion failed: (renderData), function renderClipPath, file tvgWgRenderTarget.cpp, line 111.`

3. clip beyond the rendering region (rect, for circle works fine):
```
        auto shape1 = tvg::Shape::gen();
        shape1->appendRect(500, 500, 100, 100);
        shape1->fill(255, 0, 255, 160);

        auto clipShape = tvg::Shape::gen();
        clipShape->appendRect(1000, 0, 100, 100);
        shape1->composite(std::move(clipShape), tvg::CompositeMethod::ClipPath);

        canvas->push(std::move(shape1));
```

result:
```
Validation Error

Caused by:
    In wgpuRenderPassEncoderEnd
      note: encoder = `The command encoder`
    In a set_scissor_rect command
    Scissor Rect { x: 2000, y: 0, w: 0, h: 200 } is not contained in the render target (1600, 1600, 1)

Validation Error

Caused by:
    In wgpuCommandEncoderFinish
    Command encoder is invalid
```

**Comment by SergeyLebedkin** [comment]
> @SergeyLebedkin some issues:

> 1. clipping a shape that is later used as a clip
> clip1->composite(std::move(clip2), tvg::CompositeMethod::ClipPath); **//ignored in wg?**

Thanx for a cache, must be handled recursive clipping 

> 2. please double check SceneClipper.cpp example (probably another pr?):

Yes. It will be implemented in anorher PR. Current PR cover clipping _shape with shape_ and _scene with shape_, but **not** _shape with scene_.
WIl be implemented later

> 3. clip beyond the rendering region (rect, for circle works fine):
>     Scissor Rect { x: 2000, y: 0, w: 0, h: 200 } is not contained in the render target (1600, 1600, 1)

In a case, where shape and clipping rect is not intersected, width == 0. WebGPU raise an error in this case. Need to be handled

**Comment by SergeyLebedkin** [comment]
@mgrudzinska 
Please, re-check 1 and 3 comment.
Ignore 2-nd

File: src/renderer/wg_engine/tvgWgPipelines.cpp [hunk 1]
```diff
@@ -317,7 +388,7 @@ void WgPipelineComposeBlend::initialize(WGPUDevice device)
     };
 
     // sheder source and labels
```
Comment (mgrudzinska) [inline src/renderer/wg_engine/tvgWgPipelines.cpp:390]: typo (in other places as well): sheder -> shader

**Comment by mgrudzinska** [comment]
@SergeyLebedkin I confirm it works now, didn't find any other issues

File: src/renderer/wg_engine/tvgWgPipelines.cpp (hunk 1 above)
Comment (SergeyLebedkin) [reply src/renderer/wg_engine/tvgWgPipelines.cpp]: Fixed

**Pull request merged**